tests/unit/test_charm.py

```diff
@@ -1473,3 +1478,299 @@ def test_given_gocert_available_and_not_initialized_when_configure_then_admin_us
+                                id=1, csr=str(csr), certificate_chain=[str(new_cert), str(ca)]
+                            )
+                        ]
+                    ),
+                },
+            ),
+        ):
+            context.run(Event("update-status"), state)
+        mock_set_relation_certificate.assert_called_once()
+
+    @patch(f"{TLS_LIB_PATH}.TLSCertificatesProvidesV4.get_issued_certificates")
+    @patch(f"{TLS_LIB_PATH}.TLSCertificatesProvidesV4.set_relation_certificate")
+    @patch(f"{TLS_LIB_PATH}.TLSCertificatesProvidesV4.get_certificate_requests")
+    def test_given_tls_requirers_when_certificate_revoked_in_gocert_when_configure_then_certificate_revoked(
+        self,
+        mock_get_certificate_requests,
+        mock_set_relation_certificate,
+        mock_get_issued_certificates,
+        context,
+    ):
+        state = State(
+            storage=[Storage(name="config"), Storage(name="database")],
+            containers=[Container(name="gocert", can_connect=True)],
+            networks={"juju-info": Network.default()},
+            leader=True,
+            relations=[Relation(relation_id=1, endpoint=CERTIFICATE_PROVIDER_RELATION_NAME)],
+            secrets=[
+                Secret(
+                    id="1",
+                    label=GOCERT_LOGIN_SECRET_LABEL,
+                    contents={
+                        0: {"username": "hello", "password": "world", "token": "test-token"}
+                    },
+                    owner="app",
+                )
+            ],
+        )
+        ca_pk = generate_private_key()
+        ca = generate_ca(ca_pk, 365, "me")
+        csr = generate_csr(private_key=generate_private_key(), common_name="gocert.com")
+        old_cert = generate_certificate(csr, ca, ca_pk, 365)
+        mock_get_certificate_requests.return_value = [
+            RequirerCSR(
+                relation_id=1,
+                certificate_signing_request=csr,
+            )
+        ]
+        mock_get_issued_certificates.return_value = [
+            ProviderCertificate(
+                relation_id=1,
+                certificate_signing_request=csr,
+                certificate=old_cert,
+                ca=ca,
+                chain=[old_cert, ca],
+            )
+        ]
+        with patch(
+            "gocert.GoCert.__new__",
+            return_value=Mock(
+                **{
+                    "is_api_available.return_value": True,
+                    "is_initialized.return_value": True,
+                    "token_is_valid.return_value": True,
+                    "get_certificate_requests_table.return_value": CertificateRequests(
+                        rows=[CertificateRequest(id=1, csr=str(csr), certificate_chain="rejected")]
+                    ),
+                },
+            ),
+        ):
+            context.run(Event("update-status"), state)
+        mock_set_relation_certificate.assert_called_once()
```

Comment:
Can we do something like the following? I think this applies to most asserts here, but this one is particularly interesting.
```suggestion
        mock_set_relation_certificate.assert_called_once_with(
            ProviderCertificate(
                relation_id=1,
                certificate_signing_request=csr,
                certificate=old_cert,
                ca=ca,
                chain=[old_cert, ca],
                revoked=True,
            )
        )
```

Reply:
I this to do this one but it doesn't work for some reason, so I gave up. It just fails even if the calls and the arguments are byte to byte identical. My hat's off to you if you can figure it out.